File: app/src/test/java/org/alkaline/taskbrain/ui/currentnote/ResolveNoteIdsTest.kt
package org.alkaline.taskbrain.ui.currentnote

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

/**
 * Tests for the resolveNoteIds package-level function.
 * Verifies save-time deduplication of noteIds after splits, merges, and pastes.
 */
class ResolveNoteIdsTest {

    @Test
    fun `unique noteIds pass through unchanged`() {
        val result = resolveNoteIds(
            contentLines = listOf("Line A", "Line B", "Line C"),
            lineNoteIds = listOf(listOf("id1"), listOf("id2"), listOf("id3"))
        )

        assertEquals("id1", result[0].noteId)
        assertEquals("id2", result[1].noteId)
        assertEquals("id3", result[2].noteId)
    }

    @Test
    fun `lines without noteIds get null`() {
        val result = resolveNoteIds(
            contentLines = listOf("Line A", "New line", "Line C"),
            lineNoteIds = listOf(listOf("id1"), emptyList(), listOf("id3"))
        )

        assertEquals("id1", result[0].noteId)
        assertNull(result[1].noteId)
        assertEquals("id3", result[2].noteId)
    }

    @Test
    fun `duplicate noteIds resolved by longest content`() {
        // After a split: "Hello World" -> "Hello" + " World"
        // Both claim "id1", but " World" has more content (excluding prefix)
        val result = resolveNoteIds(
            contentLines = listOf("He", "Hello World"),
            lineNoteIds = listOf(listOf("id1"), listOf("id1"))
        )

        assertNull(result[0].noteId) // shorter content loses
        assertEquals("id1", result[1].noteId) // longer content wins
    }

    @Test
    fun `duplicate noteIds with prefix - content length excludes prefix`() {
        // Tab-indented line has less actual content but more total chars
        val result = resolveNoteIds(
            contentLines = listOf("\tShort", "Much longer content here"),
            lineNoteIds = listOf(listOf("id1"), listOf("id1"))
        )

        // "\tShort" has content "Short" (5 chars)
        // "Much longer content here" has content "Much longer content here" (24 chars)
        assertNull(result[0].noteId)
        assertEquals("id1", result[1].noteId)
    }

    @Test
    fun `merge noteIds - first (primary) noteId is used`() {
        // After merging lines with noteIds ["idA", "idB"], primary is "idA"
        val result = resolveNoteIds(
            contentLines = listOf("Merged content"),
            lineNoteIds = listOf(listOf("idA", "idB"))
        )

        assertEquals("idA", result[0].noteId)
    }

    @Test
    fun `merge noteIds deduplicated when same primary appears on multiple lines`() {
        // Line 0 merged from ["idA", "idB"], line 1 also claims "idA"
        // Line 0 has longer content, so it wins "idA"
        val result = resolveNoteIds(
            contentLines = listOf("Long merged content", "Short"),
            lineNoteIds = listOf(listOf("idA", "idB"), listOf("idA"))
        )

        assertEquals("idA", result[0].noteId) // longer content wins
        assertNull(result[1].noteId) // lost the claim
    }

    @Test
    fun `empty contentLines and lineNoteIds returns empty`() {
        val result = resolveNoteIds(emptyList(), emptyList())
        assertEquals(0, result.size)
    }

    @Test
    fun `lineNoteIds shorter than contentLines - extra lines get null`() {
        val result = resolveNoteIds(
            contentLines = listOf("Line A", "Line B", "Line C"),
            lineNoteIds = listOf(listOf("id1"))
        )

        assertEquals("id1", result[0].noteId)
        assertNull(result[1].noteId)
        assertNull(result[2].noteId)
    }

    @Test
    fun `content preserved in output NoteLines`() {
        val result = resolveNoteIds(
            contentLines = listOf("\t- Item 1", "Plain text"),
            lineNoteIds = listOf(listOf("id1"), listOf("id2"))
        )

        assertEquals("\t- Item 1", result[0].content)
        assertEquals("Plain text", result[1].content)
    }

    @Test
    fun `all lines with empty noteIds all get null`() {
        val result = resolveNoteIds(
            contentLines = listOf("A", "B", "C"),
            lineNoteIds = listOf(emptyList(), emptyList(), emptyList())
        )

        assertNull(result[0].noteId)
        assertNull(result[1].noteId)
        assertNull(result[2].noteId)
    }

    @Test
    fun `equal content length - one wins deterministically`() {
        // When two lines claim the same noteId with identical content length,
        // maxByOrNull picks one deterministically (first with max value)
        val result = resolveNoteIds(
            contentLines = listOf("ABCD", "EFGH"),
            lineNoteIds = listOf(listOf("id1"), listOf("id1"))
        )

        // Both have 4 chars content. Exactly one should win
        val winners = result.filter { it.noteId == "id1" }
        assertEquals(1, winners.size)
    }
}

File: web/src/__tests__/editor/resolveNoteIds.test.ts
import { describe, it, expect } from 'vitest'
import { resolveNoteIds } from '@/editor/resolveNoteIds'

describe('resolveNoteIds', () => {
  it('passes through unique noteIds unchanged', () => {
    const result = resolveNoteIds(
      ['Line A', 'Line B', 'Line C'],
      [['id1'], ['id2'], ['id3']],
    )
    expect(result[0]!.noteId).toBe('id1')
    expect(result[1]!.noteId).toBe('id2')
    expect(result[2]!.noteId).toBe('id3')
  })

  it('gives null to lines without noteIds', () => {
    const result = resolveNoteIds(
      ['Line A', 'New line', 'Line C'],
      [['id1'], [], ['id3']],
    )
    expect(result[0]!.noteId).toBe('id1')
    expect(result[1]!.noteId).toBeNull()
    expect(result[2]!.noteId).toBe('id3')
  })

  it('resolves duplicate noteIds by longest content', () => {
    const result = resolveNoteIds(
      ['He', 'Hello World'],
      [['id1'], ['id1']],
    )
    expect(result[0]!.noteId).toBeNull()
    expect(result[1]!.noteId).toBe('id1')
  })

  it('excludes prefix when computing content length', () => {
    const result = resolveNoteIds(
      ['\tShort', 'Much longer content here'],
      [['id1'], ['id1']],
    )
    // "\tShort" content = "Short" (5 chars)
    // "Much longer content here" content = 24 chars
    expect(result[0]!.noteId).toBeNull()
    expect(result[1]!.noteId).toBe('id1')
  })

  it('uses first (primary) noteId from merges', () => {
    const result = resolveNoteIds(
      ['Merged content'],
      [['idA', 'idB']],
    )
    expect(result[0]!.noteId).toBe('idA')
  })

  it('deduplicates when merged primary appears on multiple lines', () => {
    const result = resolveNoteIds(
      ['Long merged content', 'Short'],
      [['idA', 'idB'], ['idA']],
    )
    expect(result[0]!.noteId).toBe('idA')
    expect(result[1]!.noteId).toBeNull()
  })

  it('returns empty for empty inputs', () => {
    const result = resolveNoteIds([], [])
    expect(result).toHaveLength(0)
  })

  it('handles lineNoteIds shorter than contentLines', () => {
    const result = resolveNoteIds(
      ['Line A', 'Line B', 'Line C'],
      [['id1']],
    )
    expect(result[0]!.noteId).toBe('id1')
    expect(result[1]!.noteId).toBeNull()
    expect(result[2]!.noteId).toBeNull()
  })

  it('preserves content in output', () => {
    const result = resolveNoteIds(
      ['\t- Item 1', 'Plain text'],
      [['id1'], ['id2']],
    )
    expect(result[0]!.content).toBe('\t- Item 1')
    expect(result[1]!.content).toBe('Plain text')
  })

  it('preserves noteIds through indentation changes', () => {
    // This is the key bug fix: indenting a line changes its text but
    // the editor preserves the noteId. The save should use the editor's
    // noteIds, not re-match by content.
    const result = resolveNoteIds(
      ['\t☐ Todo 2'],  // indented version
      [['id1']],        // noteId preserved by editor
    )
    expect(result[0]!.noteId).toBe('id1')
  })
})
